File: app/src/main/java/com/shentu/wallpaper/mvp/ui/widget/BaseRelativeLayout.java
package com.shentu.wallpaper.mvp.ui.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.RelativeLayout;

import butterknife.ButterKnife;

public abstract class BaseRelativeLayout extends RelativeLayout {
    public BaseRelativeLayout(Context context) {
        this(context, null);
    }

    public BaseRelativeLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        if (getLayoutId() != -1) {
            LayoutParams lp = new LayoutParams(-2, -2);
            lp.addRule(RelativeLayout.CENTER_IN_PARENT);
            addView(LayoutInflater.from(context).inflate(getLayoutId(), this, false), lp);
            ButterKnife.bind(this);
        }
        initView(context, attrs);
    }

    public int dp2px(final float dpValue) {
        final float scale = getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    public int px2dp(final float pxValue) {
        final float scale = getContext().getResources().getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }

    public int sp2px(final float spValue) {
        final float fontScale = getContext().getResources().getDisplayMetrics().scaledDensity;
        return (int) (spValue * fontScale + 0.5f);
    }

    public int px2sp(final float pxValue) {
        final float fontScale = getContext().getResources().getDisplayMetrics().scaledDensity;
        return (int) (pxValue / fontScale + 0.5f);
    }


    protected abstract int getLayoutId();

    protected abstract void initView(Context context, AttributeSet attrs);
}

File: app/src/main/java/com/shentu/wallpaper/mvp/ui/widget/BaseRelativeLayout.kt
package com.shentu.wallpaper.mvp.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.RelativeLayout
import butterknife.ButterKnife

abstract class BaseRelativeLayout @JvmOverloads constructor(context: Context , attrs: AttributeSet?,defStyleAttr:Int=-1) :RelativeLayout(context,attrs,defStyleAttr){

    fun dp2px(dpValue: Float): Int {
        val scale = resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    fun px2dp(pxValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    fun sp2px(spValue: Float): Int {
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (spValue * fontScale + 0.5f).toInt()
    }

    fun px2sp(pxValue: Float): Int {
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (pxValue / fontScale + 0.5f).toInt()
    }
}